pip-lock/main.ts

```diff
@@ -0,0 +1,392 @@
+            // Which means current would contain:
+            // `["install", "Django-4.1", "asgiref-3.5.2", "backports.zoneinfo-0.2.1", "sqlparse-0.4.2"]`
+            return {
+                state: 'end',
+                name: null,
+                version: null,
+            };
+        },
+    };
+
+    // This loop walks the lines of output from the pip dry run.
+    // We process each line we have a defined action for (in the ParseMap above),
+    // and skip ones that we don't. This could include console output with download
+    // progress indicators, and other similar artifacts that we will essentially drop.
+    let lastOutput: ParseState = {state: 'initial', name: null, version: null};
+    for(let i = 0; i < unpacked.length; i++) {
+        const trimmed = unpacked[i].trim();
+        const current: string[] = trimmed.split(" ");
+        if(!current.length)
+            throw new Error("invalid input provided");
+
+        const key = current[0];
+        const action: NullableParseStep = actions[key];
+
+        if(!action)
+            continue;
+
+        lastOutput = action(current.slice(1), lastOutput);
+    }
+
+
+    return res;
+}
+
+// We will essentially run pip with the provided arguments, parse the package output, and return the package list.
+async function checkAndRunPip(args, pipPath: string): Promise<PackageList> {
+    let packages: PackageEntry[] = [];
+
+    const pipVersion = await getVersion(pipPath);
+    if(pipVersion.major < 22 || pipVersion.minor < 2 || pipVersion.fix < 2) {
+        throw new Error(
+            `pip version ${pipVersion.major}.${pipVersion.minor}.${pipVersion.fix} found is too old, please upgrade to at least 22.2.2`);
+    }
+
+    const proc = Deno.run({
+        cmd: [pipPath, 'install', '-r', args, '--dry-run'],
+        stdout: 'piped',
+        stderr: 'piped'
+    });
+
+    try {
+        const rawOutput = await proc.output();
+        const textOutput = new TextDecoder().decode(rawOutput);
+
+        packages = parseOutput(textOutput);
+    } finally {
+        proc.close();
+    }
+
+    return {
+        package_type: "pypi",
+        packages: packages,
+    }
+}
+
+//Output strict requirements.txt
+function outputStrict(packageList: [object]) {
+    let strictList = [];
+    for(const element of packageList.packages) {
+	strictList.push(`${element.name}==${element.version}`);
+    }
+    console.log(strictList.join("\n"));
+}
+
+
+
+
+
+const pipPath = await findPip();
+if(!pipPath) {
+    console.error("[phylum] Unable to find pip - please ensure it is available in the current environment, or provide");
+    console.error("[phylum] via the `PHYLUM_PIP_PATH` environment variable.");
+    Deno.exit(1);
+}
+
+//Parse arguments
+const parsedArgs = parse(Deno.args, {"boolean": true});
+
+// Perform pip dry run install to gather relevant package list.
+const packageList = await checkAndRunPip(parsedArgs._[0], pipPath);
```

Comment:
The argument parsing here could use a bit more validation. For example, it would be good to validate is the length of `parsedArgs._`:

```suggestion
//Parse arguments
const parsedArgs = parse(Deno.args, {"boolean": true});

if(parsedArgs._ === undefined || parsedArgs._.length !== 1) {
    // TODO: Print usage here
    Deno.exit(1)
}

// Perform pip dry run install to gather relevant package list.
const lockfile = parsedArgs._[0];
const packageList = await checkAndRunPip(lockfile, pipPath);
```

Another option is to use the [`"unknown"`](https://deno.land/std@0.150.0/flags/mod.ts?s=ParseOptions#prop_unknown) function to reject unknown arguments.

Additionally, adding a `--help` argument to print usage would be great.